ModuleInstall/ExtensionManager.php

```diff
@@ -0,0 +1,206 @@
+     *
+     * @param string $extension Name of Extension i.e. 'Language'
+     * @param string $targetFileName Name of target file
+     * @param string $filter To filter file names such as language prefixes
+     * @param bool $applicationOnly Whether or not to only compile application extensions
+     * @return void
+     */
+    public static function compileExtensionFiles(
+        $extension,
+        $targetFileName,
+        $filter = '',
+        $applicationOnly = false
+    ) {
+        static::initialise();
+
+        if ($extension === 'Language' && strpos($targetFileName, $filter) !== 0) {
+            $targetFileName = $filter . $targetFileName;
+        }
+
+        if (!$applicationOnly) {
+            static::compileModuleExtensions($extension, $targetFileName, $filter);
+        }
+
+        static::compileApplicationExtensions($extension, $targetFileName, $filter);
+    }
+
+    /**
+     * @param $extension
+     * @param $targetFileName
+     * @param string $filter
+     * @return void
+     */
+    public static function compileApplicationExtensions(
+        $extension,
+        $targetFileName,
+        $filter = ''
+    ) {
+        static::$logger->{'debug'}("Merging application files for $targetFileName in $extension");
+
+        $extensionContents = '<?php' . PHP_EOL . '// WARNING: The contents of this file are auto-generated' . PHP_EOL;
+        $extPath = "application/Ext/$extension";
+        $moduleInstall  = "custom/Extension/$extPath";
+        $shouldSave = false;
+
+        if (is_dir($moduleInstall)) {
+            $dir = dir($moduleInstall);
+
+            while ($entry = $dir->read()) {
+                $shouldSave = true;
+                if ($entry === '.' || $entry === '..' || strtolower(substr($entry, -4)) !== '.php') {
+                    continue;
+                }
+
+                if (!is_file("$moduleInstall/$entry")) {
+                    continue;
+                }
+
+                if (!empty($filter) && substr_count($entry, $filter) <= 0) {
+                    continue;
+                }
+
+                $file = file_get_contents("$moduleInstall/$entry");
+                $extensionContents .= PHP_EOL . static::removePhpTagsFromString($file);
+            }
+        }
+
+        $extensionContents .= "\n?>";
+
+        if ($shouldSave) {
+            if (!file_exists("custom/$extPath")) {
+                mkdir_recursive("custom/$extPath", true);
+            }
+
+            $out = sugar_fopen("custom/$extPath/$targetFileName", 'wb');
+            fwrite($out, $extensionContents);
+            fclose($out);
+
+            return;
+        }
+
+        if (file_exists("custom/$extPath/$targetFileName")) {
+            unlink("custom/$extPath/$targetFileName");
+        }
+    }
+
+    /**
+     * @param $extension
+     * @param $targetFileName
+     * @param string $filter
+     * @return void
+     */
+    public static function compileModuleExtensions(
```

Comment:
This method is very long.

Reply:
Same as above, will refactor it shortly.

Reply:
Little better?